### DISCUSSION TITLE
Generating a Route in a Macro

### DISCUSSION THREAD

**Discussion by jafioti** [rwf2/Rocket#2058] [Questions] [answered]
Hi,
I would like to generate a route for a function in a macro. For instance, I have a function:
```
fn hello() {
     println!("Hello");
}
```
And I would like to generate a route for it:
```
#[get("/hello")]
fn hello_route() {
     hello()
}
```

How would I go about doing this? I currently have
```
#[macro_export]
macro_rules! generate_functions {
    ($vis:vis fn $name:ident($($arg:ident: $arg_type:ty)*) $code:block) => {
        ::paste::paste! {
            $vis fn $name($($arg: $arg_type)*) $code

            #[get("/" $name)]
            $vis fn [<$name _route>]($($arg: $arg_type)*) {
                $name($($arg)*)
            }
        }
    };
}
```
But that doesn't work because it doesn't properly set the route path in the #[get()] macro.

**Comment by the10thWiz** [comment]
You probably want to use the `concat!` macro to combine the path into a single literal.

```#[get(concat!("/", $name)]```

That being said, why do you want to do this? Rocket routes are still functions you can call, so

```
#[get("/hello")]
fn hello() -> &str {
  "Hello world!"
}

fn other() {
  // This is perfectly valid
  hello();
}
```

Rocket already internally generates a wrapper that just calls your function.

**Comment by jafioti** [comment, marked as answer]
I ended up having to use a proc macro. I packaged it up in a crate here: https://github.com/Sidekick-AI/retrofit
